crates/sui-core/src/consensus_handler.rs

```diff
@@ -829,6 +858,104 @@ impl ConsensusCommitInfo {
+    }
+
+    pub async fn handle_consensus_transactions(&self, parsed_transactions: Vec<ParsedTransaction>) {
+        let mut pending_consensus_transactions = vec![];
+        let executable_transactions: Vec<_> = parsed_transactions
+            .into_iter()
+            .filter_map(|parsed| {
+                // TODO(fastpath): unlock rejected transactions.
+                // TODO(fastpath): maybe avoid parsing blocks twice between commit and transaction handling?
+                if parsed.rejected {
+                    self.metrics
+                        .consensus_transaction_handler_processed
+                        .with_label_values(&["rejected"])
+                        .inc();
+                    return None;
+                }
+                self.metrics
+                    .consensus_transaction_handler_processed
+                    .with_label_values(&["certified"])
+                    .inc();
+                match &parsed.transaction.kind {
+                    ConsensusTransactionKind::UserTransaction(tx) => {
+                        if tx.contains_shared_object() {
+                            return None;
+                        }
+                        pending_consensus_transactions.push(parsed.transaction.clone());
+                        let tx = VerifiedTransaction::new_from_verified(*tx.clone());
+                        Some(VerifiedExecutableTransaction::new_from_consensus(
+                            tx,
+                            self.epoch_store.epoch(),
+                            parsed.round,
+                            parsed.authority,
+                            parsed.transaction_index,
+                        ))
+                    }
+                    _ => None,
+                }
+            })
+            .collect();
+
+        if pending_consensus_transactions.is_empty() {
+            return;
+        }
+        {
+            let reconfig_state = self.epoch_store.get_reconfig_state_read_lock_guard();
+            // Stop executing fastpath transactions when epoch change starts.
+            if !reconfig_state.should_accept_user_certs() {
+                return;
+            }
+            // Otherwise, try to ensure the certified transactions get into consensus before epoch change.
+            // TODO(fastpath): avoid race with removals in consensus adapter, by waiting to handle commit after
+            // all blocks in the commit are processed via the transaction handler. Other kinds of races need to be
+            // avoided as well. Or we can track pending consensus transactions inside consensus instead.
+            self.epoch_store
+                .insert_pending_consensus_transactions(
+                    &pending_consensus_transactions,
+                    Some(&reconfig_state),
+                )
+                .unwrap_or_else(|e| {
+                    panic!("Failed to insert pending consensus transactions: {}", e)
+                });
```

Comment:
Is it guaranteed (I guess not) at this point that only non already executed transactions will go through here , or it would be possible to receive duplicates? Just thinking whether we want to do any sanity check